### PR TITLE
[21160] Remove rails work_packages views and update specs

### PR DESCRIPTION
**Pull request by oliverguenther**
### TODOS
- [x] Test cukes on their own and remove outdated ones
- [x] Write feature specs for full screen create
### Bugs

(Extend this list with bugs you find)
- [x] Edit work package used to the old view. `/work_packages/:id/edit`  must be defined in angular and should route to the split screen in force edit mode.
- [X] ~~Copy redirects to the `work_package_path` (`/work_packages/:id`) which shows an empty page - the angular router doesn't chime in here.~~ Fixed with https://github.com/oliverguenther/openproject/commit/3ad9ff3a80567d4bb04c3b1220ae3688f22e5f30
- [x] ~~Duplicate work packages used the old view. It 'works' in the sense that no error is thrown but no values are taken over.~~ I have removed the duplicate option for now. The code for it still exists, since we will likely implement it soon enough.

Relevant work package: https://community.openproject.org/work_packages/21160/activity

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://oliverguenther-features-16364-remove-rails-views.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-features-16364-remove-rails-views.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by oliverguenther** [comment]
Removed cukes:
- create.feature: Uses no JS
- copy_with_watchers.feature: Copy feature not implemented
- preview.feature: No preview available

**Comment by ulferts** [comment]
The removed actions are still referenced in the permissions definition, e.g.:
- https://github.com/opf/openproject/blob/dev/config/initializers/permissions.rb#L121
- https://github.com/opf/openproject/blob/dev/config/initializers/permissions.rb#L132
- https://github.com/opf/openproject/blob/dev/config/initializers/permissions.rb#L140

(not necessarily complete)

**Comment by ulferts** [comment]
There are a whole lot of methods in https://github.com/opf/openproject/blob/dev/app/helpers/work_packages_helper.rb which are no longer necessary I think. Of course, this will not pose immediate problems as this is now code that is no longer referenced, but It will slow down development.

File: config/routes.rb [hunk 1]
```diff
@@ -301,9 +301,6 @@
       resources :calendar, controller: 'calendars', only: [:index]
     end
 
-    # Prevent the /*state route from overriding Rails' new
-    # view which is being used for e.g. duplicating work packages.
-    resources :work_packages, only: [:new]
 
     resources :work_packages, only: [:create] do
```
Comment (ulferts) [inline config/routes.rb]: The `:create` action no longer exists, right?

Comment (ulferts) [inline config/routes.rb]: There are other references to no longer existing actions in the routes, e.g.:
https://github.com/oliverguenther/openproject/blob/features/16364-remove_rails_views/config/routes.rb#L428

File: spec/views/work_package/show.html.erb_spec.rb [hunk 2]
```diff
@@ -37,15 +37,6 @@
...
-  it 'renders correct image paths in journal entries' do
-    work_package.add_journal(work_package.author, 'bar !foo.jpg! bar')
-    work_package.attachments << attachment
-    work_package.save
-    render 'history', work_package: work_package, journals: work_package.journals
-    expect(rendered)
-      .to have_selector "img[src='/attachments/#{attachment.id}/download']"
-  end
-
   context 'watchers list is sorted alphabeticaly' do
```
Comment (ulferts) [inline spec/views/work_package/show.html.erb_spec.rb:40]: If it is not to much effort, I would move this to somewhere else. It was never a test for work_packages/show to begin with and now this is looking really weird.

File: config/initializers/permissions.rb [hunk 3]
```diff
@@ -127,7 +127,7 @@
                    require: :loggedin
 
     wpt.permission :edit_work_packages,
-                   { issues: [:edit, :update, :update_form],
+                   { issues: [:edit, :update],
                      :'work_packages/bulk' => [:edit, :update],
                      work_packages: [:edit, :update, :new_type,
                                      :preview, :quoted],
```
Comment (ulferts) [inline config/initializers/permissions.rb:133]: What about those here:
`:new_type, :preview, :quoted`
They are not used in the API v2 so they should be removable.

File: config/routes.rb [hunk 4]
```diff
@@ -427,13 +420,9 @@
     get '/bulk' => 'bulk#destroy'
   end
 
-  resources :work_packages, only: [:update, :index] do
-    get :new_type, on: :member
-
+  resources :work_packages, only: [:index] do
     get :column_data, on: :collection # TODO move to API
 
-    resources :relations, controller: 'work_package_relations', only: [:create, :destroy]
```
Comment (ulferts) [inline config/routes.rb:435]: I think the `WorkPackageRelationsController` can be removed as well.

File: frontend/app/work_packages/controllers/menus/work-package-context-menu-controller.js [hunk 5]
```diff
@@ -99,6 +99,11 @@ module.exports = function(
     return WorkPackagesTableHelper.getWorkPackagesFromRows(selectedRows);
   }
 
+  function getSelectedWorkPackageId() {
+    var selected = getSelectedWorkPackages()[0];
+    return selected && selected.id;
+  }
+
```
Comment (ulferts) [inline frontend/app/work_packages/controllers/menus/work-package-context-menu-controller.js]: You removed this in https://github.com/opf/openproject/pull/3835/files#diff-ab5294bb05227591715f26738a3b98aeL90

File: config/initializers/permissions.rb (hunk 3 above)
Comment (oliverguenther) [inline config/initializers/permissions.rb:133]: ~~Yup, will do.~~ Done.

File: spec/features/work_packages/edit_work_package_spec.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  it 'shows the work package in edit mode' do
+    subject = page.find("#inplace-edit--write-value--subject")
+    expect(subject.value).to eq(work_package.subject)
+
+    subject.set new_subject
+    find('#work-packages--edit-actions-save').click
+
+    subject_field.expect_state_text(subject)
+  end
```
Comment (ulferts) [inline spec/features/work_packages/edit_work_package_spec.rb:29]: There is a [Page Object](https://github.com/opf/openproject/blob/dev/spec/support/pages/full_work_package.rb) that might help here in addition to the one already used.

**Comment by ulferts** [comment]
Nice added vs. removed ratio:

image

More of those, please.

**Comment by oliverguenther** [comment]
> Nice added vs. removed ratio:

If only the ratio of added vs. removed tests would not be so highly identical to the ratio of removed lines. Oh well, can't have them all. :grimacing:

File: spec/views/work_package/show.html.erb_spec.rb (hunk 2 above)
Comment (oliverguenther) [inline spec/views/work_package/show.html.erb_spec.rb:40]: Good idea. Moved to its own `watchers/watchers.html.erb_spec`.

File: config/routes.rb (hunk 4 above)
Comment (oliverguenther) [inline config/routes.rb:435]: Done in updated https://github.com/oliverguenther/openproject/commit/764d54dc698ae0649a5354d8103386dbe6fd67ff

File: spec/features/work_packages/edit_work_package_spec.rb (hunk 6 above)
Comment (oliverguenther) [inline spec/features/work_packages/edit_work_package_spec.rb:29]: That one only works for the fullscreen, doesn't it? 

The edit route currently leads to the split screen force edit mode. I can extend the test to expand the edit mode to fullscreen, if that's your intention?

File: spec/views/work_package/show.html.erb_spec.rb (hunk 2 above)
Comment (ulferts) [inline spec/views/work_package/show.html.erb_spec.rb:40]: Thanks

**Comment by oliverguenther** [comment]
The last cuke job cannot be restarted, but this is ready to merge from my side.

**Comment by ulferts** [comment]
I'm merging this now and we'll see what our internal CI has to say.

**Pull request merged**